### ISSUE TITLE
[Q] How to use `wandb.init` to change project settings?

### ISSUE THREAD

**Issue by why-in-Shanghaitech** [wandb/wandb#5272]
In the docs: https://docs.wandb.ai/guides/app/pages/project-page#project-defaults, it says

> These settings can also be specified by passing arguments to [wandb.init](https://docs.wandb.ai/ref/python/init).

But how? Specifically, how to set project privacy through `wandb.init`?

**Comment by thanos-wandb** [comment]
Hi @why-in-Shanghaitech thanks for reporting this, I will create a documentation update request as project permissions settings don't seem to be an option from the `wandb.init()` call. 

However, you could use the API to create a project before creating a new run in it as follows:

```
import wandb

api = wandb.Api()
api.client.execute(
    api.CREATE_PROJECT,
    {
        'entityName': 'your-entity-name',
        'name': 'your-project-name',
        'access': 'PRIVATE',
    },
)
```

You could have access set to `USER_READ` (for `Public` project) or `USER_WRITE` (for `Open` project). I hope this helps, please let me know if that won't work  for you.

**Comment by why-in-Shanghaitech** [comment]
Thank you very much for the reply! Unfortunately, it does not work in version `0.12.9`, since it raises `AttributeError: 'Api' object has no attribute 'CREATE_PROJECT'`. Though it works in the latest version `0.14.0`, I cannot log results in an environment without the Internet (even with environment variable `WANDB_MODE` as `dryrun` and setting `mode="offline"` in `wandb.init`). Version `0.14.0` will try to access `localhost`, which is impossible in my case.

```
   6    File "<path>/wandb/sdk/lib/sock_client.py", line 102, in connect
   7:     s.connect(("localhost", port))
   8  ConnectionRefusedError: [Errno 111] Connection refused
```

Since this is a new problem, I'm closing this issue. I have no plan to create a new issue because currently I'll just stick to version `0.12.9`.

**Comment by thanos-wandb** [comment]
Hi @why-in-Shanghaitech thanks for the confirmation that this worked for you on `v0.14.0`. 

I agree that the new issue would be best to go in a new thread to be tracked but since you've mentioned you aren't planning on making one, I was wondering if we could quickly troubleshoot this one, are you running Sweeps or simple runs? The `dryrun` has been deprecated in our most recent versions, so you will need to either set it from the env `WANDB_MODE=offline`, or from` wandb.init(.., mode="offline")`. Do you have any other WANDB_ environment variables set, and are you logging runs to our public cloud instance wandb.ai or to a Local W&B instance?

**Comment by why-in-Shanghaitech** [comment]
Thanks for the reply! I try to change the environment variable, but it still doesn't work. I'll attach the full error message below:

```
Traceback (most recent call last):
  File "<path>/wandb/sdk/wandb_manager.py", line 117, in _service_connect
    svc_iface._svc_connect(port=port)
  File "<path>/wandb/sdk/service/service_sock.py", line 30, in _svc_connect
    self._sock_client.connect(port=port)
  File "<path>/wandb/sdk/lib/sock_client.py", line 102, in connect
    s.connect(("localhost", port))
ConnectionRefusedError: [Errno 111] Connection refused

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<path>/wandb/sdk/wandb_init.py", line 1140, in init
    wi.setup(kwargs)
  File "<path>/wandb/sdk/wandb_init.py", line 171, in setup
    self._wl = wandb_setup.setup(settings=setup_settings)
  File "<path>/wandb/sdk/wandb_setup.py", line 327, in setup
    ret = _setup(settings=settings)
  File "<path>/wandb/sdk/wandb_setup.py", line 320, in _setup
    wl = _WandbSetup(settings=settings)
  File "<path>/wandb/sdk/wandb_setup.py", line 303, in __init__
    _WandbSetup._instance = _WandbSetup__WandbSetup(settings=settings, pid=pid)
  File "<path>/wandb/sdk/wandb_setup.py", line 114, in __init__
    self._setup()
  File "<path>/wandb/sdk/wandb_setup.py", line 250, in _setup
    self._setup_manager()
  File "<path>/wandb/sdk/wandb_setup.py", line 277, in _setup_manager
    self._manager = wandb_manager._Manager(settings=self._settings)
  File "<path>/wandb/sdk/wandb_manager.py", line 164, in __init__
    sentry_reraise(e, delay=True)
  File "<path>/wandb/util.py", line 214, in sentry_reraise
    raise exc.with_traceback(sys.exc_info()[2])
  File "<path>/wandb/sdk/wandb_manager.py", line 162, in __init__
    self._service_connect()
  File "<path>/wandb/sdk/wandb_manager.py", line 126, in _service_connect
    raise ManagerConnectionRefusedError(message)
wandb.sdk.wandb_manager.ManagerConnectionRefusedError: Connection to wandb service failed since the process is not available. 
wandb: ERROR Abnormal program exit
Traceback (most recent call last):
  File "<path>/wandb/sdk/wandb_manager.py", line 117, in _service_connect
    svc_iface._svc_connect(port=port)
  File "<path>/wandb/sdk/service/service_sock.py", line 30, in _svc_connect
    self._sock_client.connect(port=port)
  File "<path>/wandb/sdk/lib/sock_client.py", line 102, in connect
    s.connect(("localhost", port))
ConnectionRefusedError: [Errno 111] Connection refused

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<path>/wandb/sdk/wandb_init.py", line 1140, in init
    wi.setup(kwargs)
  File "<path>/wandb/sdk/wandb_init.py", line 171, in setup
    self._wl = wandb_setup.setup(settings=setup_settings)
  File "<path>/wandb/sdk/wandb_setup.py", line 327, in setup
    ret = _setup(settings=settings)
  File "<path>/wandb/sdk/wandb_setup.py", line 320, in _setup
    wl = _WandbSetup(settings=settings)
  File "<path>/wandb/sdk/wandb_setup.py", line 303, in __init__
    _WandbSetup._instance = _WandbSetup__WandbSetup(settings=settings, pid=pid)
  File "<path>/wandb/sdk/wandb_setup.py", line 114, in __init__
    self._setup()
  File "<path>/wandb/sdk/wandb_setup.py", line 250, in _setup
    self._setup_manager()
  File "<path>/wandb/sdk/wandb_setup.py", line 277, in _setup_manager
    self._manager = wandb_manager._Manager(settings=self._settings)
  File "<path>/wandb/sdk/wandb_manager.py", line 164, in __init__
    sentry_reraise(e, delay=True)
  File "<path>/wandb/util.py", line 214, in sentry_reraise
    raise exc.with_traceback(sys.exc_info()[2])
  File "<path>/wandb/sdk/wandb_manager.py", line 162, in __init__
    self._service_connect()
  File "<path>/wandb/sdk/wandb_manager.py", line 126, in _service_connect
    raise ManagerConnectionRefusedError(message)
wandb.sdk.wandb_manager.ManagerConnectionRefusedError: Connection to wandb service failed since the process is not available. 

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<path>/original-test/example/model.py", line 43, in <module>
    optimized_params = nni.get_next_parameter()
  File "<path>/nni/nni/trial.py", line 64, in get_next_parameter
    _params = platform.get_next_parameter()
  File "<path>/nni/nni/runtime/platform/local.py", line 60, in get_next_parameter
    wandb_init(params)
  File "<path>/nni/nni/runtime/platform/local.py", line 94, in wandb_init
    wandb.init(
  File "<path>/wandb/sdk/wandb_init.py", line 1181, in init
    raise Exception("problem") from error_seen
Exception: problem
```

1. All environment variables starts with `WANDB_` are: `WANDB_MODE` with value `offline`.
2. The code for initialization:

```python
wandb.init(
        dir="<path>",

        # set the wandb project where this run will be logged
        project="<random string>",
        name="<random string>",
        id="<random string>",
        
        # track hyperparameters and run metadata
        config=params,
        anonymous="allow",
        mode="offline"
    )
```

3. I am logging to public cloud instance wandb.ai. I am using an anonymous account.

```
wandb: Currently logged in as: anony-moose-529863. Use `wandb login --relogin` to force relogin
```

**Comment by thanos-wandb** [comment]
Hi @why-in-Shanghaitech I was wondering if you could try to disable `wandb service`, if that would help to overcome this issue. Could you please add the following lines before your `wandb.init` call? 

```
import wandb
import os

os.environ['WANDB_DISABLE_SERVICE']='True'
```

**Comment by thanos-wandb** [comment]
Hey @why-in-Shanghaitech just checking in here to see if you've tried disabling `service`, and if that worked for you? thanks!

**Comment by why-in-Shanghaitech** [comment]
Sorry for the late reply. I have to say that this issue is more complicated than I thought.

I was developing a patch for NNI with slurm: https://github.com/whyNLP/nni-slurm. The network issue only appears when I use wandb in NNI trials (submitting to slurm). If I submit a job directly to slurm, setting `WANDB_MODE='offline'` or `init(mode='offline')` is enough to make the script work.

So the good news is that this is not a common issue for most users. I haven't tried to disable wandb service on NNI. I'm quite busy with my coursework and research these days, but once I get time I'll give it a try and post the results here.

**Comment by thanos-wandb** [comment]
Thanks @why-in-Shanghaitech for developing and sharing this patch, this is great! You may also find useful this [unofficial repo](https://github.com/elyall/wandb_on_slurm) regarding wandb on Slurm. Since the original issue has been resolved, I will close this ticket for now. However, please feel free to reopen this or create a new thread once you're available to resume to this and we will be glad to keep investigating!

**Comment by gil2rok** [comment]
I had the same error message 

> wandb.sdk.wandb_manager.ManagerConnectionRefusedError: Connection to wandb service failed since the process is not available.

when using SLURM and Weights and Biases hyper-parameter sweep in parallel.

This error occurs because there is a W&B bug that tries to connect to `localhost` instead of the W&B controller that schedules which hyper-parameter configuration to run next.

I set `os.environ['WANDB_DISABLE_SERVICE']= "True"` which completely fixed the problem. This was a difficult bug to solve so please fix it in the next W&B release! Thanks.

**Comment by michaelmior** [comment]
@gil2rok Not sure how you discovered that was the solution, but I ran into the same problem when using W&B on Slurm. Thanks for the temporary fix!

**Comment by kptkin** [comment]
@gil2rok sorry for the bug and thanks a lot for finding the source of the issue. We would like to fix this issue for you so you don't need to disable the wandb service, do you think you would be able to provide us with a small reproduction, so we could fix it (allowing both wandb service and proper usage of sweeps?)

**Comment by kptkin** [comment]
@michaelmior are you also using it in combination with sweeps or do you see this error just when using wandb on Slurm?

**Comment by michaelmior** [comment]
@kptkin Only in combination with sweeps. I've been able to use wandb logging in other Slrum jobs without issue.

**Comment by kptkin** [comment]
> @kptkin Only in combination with sweeps. I've been able to use wandb logging in other Slrum jobs without issue. Could you clarify what `WANDB_DISABLE_SERVICE` does? It doesn't seem to be documented anywhere.

Thanks for the response. 
Sure, it is just some internal implementation details, but high level speaking:
-  In our default mode of operation we spin up a separate process (which we call service), this process manages all the active runs in your system and responsible for the network and file system operations. 
- When you disable this flag we revert to a different mode of operation where each run will have it's own process to manage the run. You also lose features like sharing the run across multiple processes and in general, risking a hang if the multiprocessing with other code is not done properly.
- For some additional context see here: https://docs.wandb.ai/guides/track/log/distributed-training#enable-wb-service (as of 0.16.0 it is enabled by default, need to update the documentation)

Hope this helps. @michaelmior if you could provide a small reproduction example i would be happy to resolve this issue, so you don't have to disable this service.

**Comment by michaelmior** [comment]
@kptkin I've been able to reproduce this using [litgpt](https://github.com/Lightning-AI/litgpt) with the following sweep

```yml
method: bayes
metric:
  goal: minimize
  name: val_loss
parameters:
  train.learning_rate:
    min: 0.00001
    max: 0.01
  train.min_lr:
    min: 0.000001
    max: 0.00006
  train.epochs:
    min: 1
    max: 5
  train.weight_decay:
    min: 0.0
    max: 0.05
command:
  - ${env}
  - sbatch
  - --wait
  - ./run.sh
  - --config
  - lora.yaml
  - ${args}
```

I'm using a custom dataset, but I think using any config like [this one](https://raw.githubusercontent.com/Lightning-AI/litgpt/6d04a87988ecf499b2a9ec1cfe1ad529ffb97894/config_hub/finetune/llama-2-7b/lora.yaml) will work.

My `run.sh` script is just a wrapper that sets up my environment. Ultimately it just executes `litgpt finetune lora` with all the arguments passed in.

**Comment by amacrutherford** [comment]
I have also run into this issue when running on a server (not using SLURM). `os.environ['WANDB_DISABLE_SERVICE']= "True"` seems to have fixed it. It was working fine last week and then randomly started throwing the same error as [listed previously](https://github.com/wandb/wandb/issues/5272#issuecomment-1505125344).

**Comment by DavidoF3** [comment]
I have similar problems. Iam trying resume a wandb run on a SLURM job, by running:

```
run = wandb.init(
        entity=entity,
        project=project,
        id=<ide_of_run>,
        resume="must",
    )
```
And I get the errors:

```
ConnectionRefusedError: [Errno 111] Connection refused
During handling of the above exception, another exception occurred:
wandb.sdk.wandb_manager.ManagerConnectionRefusedError: Connection to wandb service failed since the process is not available. 
```

I pass my API key via:
`export WANDB_API_KEY=` in the sbatch script

**Comment by kptkin** [comment]
@DavidoF3 i think your issue is a bit different, could be that your environment is in a bad state and we can't connect to the service process. Could  you make sure that you don't have any value set for `"WANDB_SERVICE"` env variable? also if you happen to have: `ps -ef | grep wandb-service` process could you please kill these as well?

**Comment by kptkin** [comment]
@amacrutherford if it throws this error it means you didn't actually disable service as it still tries to connect to the service process. see my comment here: https://github.com/wandb/wandb/issues/5272#issuecomment-2043405987

**Comment by DavidoF3** [comment]
@kptkin I am able to run a job successfully when the service is disabled. However, I would like to run my job with the service ON because I would like to keep track of training results from the wandb UI.

**Comment by michaelmior** [comment]
@kptkin You should be able to run the job with the service disabled and still track the results from the UI. At least, I am able to do so just fine in my jobs.

**Comment by kptkin** [comment]
@michaelmior @DavidoF3 yes service shouldn't affect your ability to track the results in the UI. It is just enable more advanced functionality to manage multiprocessing that log to the same run and other improvements

**Comment by kptkin** [comment]
> @kptkin I've been able to reproduce this using [litgpt](https://github.com/Lightning-AI/litgpt) with the following sweep
> 
> ```yaml
> method: bayes
> metric:
>   goal: minimize
>   name: val_loss
> parameters:
>   train.learning_rate:
>     min: 0.00001
>     max: 0.01
>   train.min_lr:
>     min: 0.000001
>     max: 0.00006
>   train.epochs:
>     min: 1
>     max: 5
>   train.weight_decay:
>     min: 0.0
>     max: 0.05
> command:
>   - ${env}
>   - sbatch
>   - --wait
>   - ./run.sh
>   - --config
>   - lora.yaml
>   - ${args}
> ```
> 
> I'm using a custom dataset, but I think using any config like [this one](https://raw.githubusercontent.com/Lightning-AI/litgpt/6d04a87988ecf499b2a9ec1cfe1ad529ffb97894/config_hub/finetune/llama-2-7b/lora.yaml) will work.
> 
> My `run.sh` script is just a wrapper that sets up my environment. Ultimately it just executes `litgpt finetune lora` with all the arguments passed in.

Thanks for the reproduction, i'm adding it to our log of tasks, hopefully we can solve it for you guys soon!

**Comment by DavidoF3** [comment]
@michaelmior you are right. The reason why it wasn't updating the UI is because I was running both:
```
export WANDB_MODE='offline'     -> (suggested in other issues)
export WANDB_DISABLE_SERVICE='True'
```
If you want to track results in the UI, you just need to run:
```
export WANDB_DISABLE_SERVICE='True'
```
